gc/base/standard/ConcurrentGC.cpp

```diff
@@ -95,35 +94,33 @@ extern "C" {
 
 /**
  * Concurrent mark write barrier
- * External entry point for write barrier. Calls card table function to dirty card for
- * given object which has had an object reference modified.
+ * External entry point for write barrier.
  *
  * @param destinationObject Address of object containing modified reference
  * @param storedObject The value of the modified reference
  */
 void
-J9ConcurrentWriteBarrierStore (OMR_VMThread *vmThread, omrobjectptr_t destinationObject, omrobjectptr_t storedObject)
+J9ConcurrentWriteBarrierStore(OMR_VMThread *vmThread, omrobjectptr_t destinationObject, omrobjectptr_t storedObject)
 {
 	MM_EnvironmentBase *env = MM_EnvironmentBase::getEnvironment(vmThread);
 	MM_GCExtensionsBase *extensions = env->getExtensions();
 
-	extensions->cardTable->dirtyCard(env, (omrobjectptr_t)destinationObject);
+	((MM_ConcurrentGC *)extensions->getGlobalCollector())->J9ConcurrentWriteBarrierStoreHandler(env, destinationObject, storedObject);
```

Comment:
Another approach is to specialize this API into 2 variants: pre and post. Former would be used only for SATB and latter only for IncreemtalUpdate style of CM. The hope is, caller would  know which one is active, and call approprate one. At least it seems to be the case in MM_StandardAccessBarrier::postObjectStoreImpl.

Then, each of 2 APIs would deal directly with appropriate CM class, without  a need for virtual dispatch.

I'm slightly concerned about virtual dispatch perf implications, although the truth is this is only excercised by interpreter, varius native code etc., since JIT inlines this completely. 

Regardless of perf implications , we generally have separate APIs for pre and post barriers.

Reply:
Related... We already have these even heaviers APIs for external users that should be also properly renamed (again, I'd use generic pre and post prefixes, rather implying anything about underlying technology, like Metronome, SATB, Card Dirtying etc):

J9MetronomeWriteBarrierStore and J9WriteBarrierStore

Reply:
Ah another minor one: J9 should be dropped from J9ConcurrentWriteBarrierStore.

Reply:
Addressed by recent J9 Commits - https://github.com/eclipse-openj9/openj9/pull/13046 & https://github.com/eclipse-openj9/openj9/pull/13080

Reply:
I've added `concurrentPostWriteBarrierStore` as a replacement for both `J9ConcurrentWriteBarrierStore/J9ConcurrentWriteBarrierBatchStore`. However, I'm leaving the existing APIs in case other langs/projects implement CM (as discussed)